File: Array/yanghao/java/Algorithm6_15.java
package jdbc.mvc;

import java.util.List;
import java.util.Stack;

public class Algorithm6_15<V> {
	
	public static boolean isValid(String s) {
		char[] a = s.toCharArray();
		if (a.length <= 0) {
			return false;
		}
		
		if (a.length %2 == 1) {
			return false;
		}
		Stack<Character> stack = new Stack<>();
		for(int i = 0; i < s.length(); i++) {
			char theChar = s.charAt(i);
			if (theChar == '(' || theChar == '{' || theChar == '[') {
				stack.push(theChar);
			}else {
				if (stack.isEmpty()) {
					return false;
				}
				
				char preChar = stack.peek();
				if ((preChar == '{' && theChar == '}') || (preChar == '(' && theChar == ')') || (preChar == '[' && theChar == ']')) {
					stack.pop();
				}else {
					return false;
				}
			}
		}
		return stack.isEmpty();
    }
	
	public static void main(String[] args) {
		System.out.println(isValid("{}"));
	}
}

File: Array/yanghao/kotlin/Algorithm6_15.kt
package staudy

import java.util.*

class Algorithm6_15 {
    fun isValid(s: String): Boolean {
        if (s.length % 2 == 1) return false
        val stack:Stack<Char> = Stack()
        s.toCharArray().forEach{c ->
            if (c == '{' || c == '[' || c == '('){
                stack.push(c)
            }else{
                if (stack.isEmpty()){
                    return false
                }

                val preChar = stack.peek()
                if ((preChar == '{' && c == '}') || (preChar == '(' && c == ')') || (preChar == '[' && c == ']')) stack.pop()
                else return false
            }
        }
        return stack.isEmpty()
    }

}

fun main(args: Array<String>) {
    print(Algorithm6_15().isValid("{}"))
}